src/cfi.rs

```diff
@@ -475,6 +499,520 @@ pub enum RegisterRule<'input, Endian>
+    ValExpression {
+        /// The target register's number.
+        register: u64,
+        /// The DWARF expression.
+        expression: EndianBuf<'input, Endian>,
+    },
+
+    /// The `Restore` instruction represents both `DW_CFA_restore` and
+    /// `DW_CFA_restore_extended`.
+    ///
+    /// > 11. DW_CFA_restore
+    /// >
+    /// > The DW_CFA_restore instruction takes a single operand (encoded with
+    /// > the opcode) that represents a register number. The required action is
+    /// > to change the rule for the indicated register to the rule assigned it
+    /// > by the initial_instructions in the CIE.
+    Restore {
+        /// The register to be reset.
+        register: u64,
+    },
+
+    // 6.4.2.4 Row State Instructions
+    /// > 1. DW_CFA_remember_state
+    /// >
+    /// > The DW_CFA_remember_state instruction takes no operands. The required
+    /// > action is to push the set of rules for every register onto an implicit
+    /// > stack.
+    RememberState,
+
+    /// > 2. DW_CFA_restore_state
+    /// >
+    /// > The DW_CFA_restore_state instruction takes no operands. The required
+    /// > action is to pop the set of rules off the implicit stack and place
+    /// > them in the current row.
+    RestoreState,
+
+    // 6.4.2.5 Padding Instruction
+    /// > 1. DW_CFA_nop
+    /// >
+    /// > The DW_CFA_nop instruction has no operands and no required actions. It
+    /// > is used as padding to make a CIE or FDE an appropriate size.
+    Nop,
+}
+
+const CFI_INSTRUCTION_HIGH_BITS_MASK: u8 = 0b11000000;
+const CFI_INSTRUCTION_LOW_BITS_MASK: u8 = !CFI_INSTRUCTION_HIGH_BITS_MASK;
+
+impl<'input, Endian> CallFrameInstruction<'input, Endian>
+    where Endian: Endianity
+{
+    fn parse(input: EndianBuf<'input, Endian>)
+             -> ParseResult<(EndianBuf<'input, Endian>, CallFrameInstruction<'input, Endian>)> {
+        let (rest, instruction) = try!(parse_u8e(input));
+        let high_bits = instruction & CFI_INSTRUCTION_HIGH_BITS_MASK;
+
+        if high_bits == constants::DW_CFA_advance_loc.0 {
+            let delta = instruction & CFI_INSTRUCTION_LOW_BITS_MASK;
+            return Ok((rest, CallFrameInstruction::AdvanceLoc { delta: delta as u32 }));
+        }
+
+        if high_bits == constants::DW_CFA_offset.0 {
+            let register = instruction & CFI_INSTRUCTION_LOW_BITS_MASK;
+            let (rest, offset) = try!(parse_unsigned_lebe(rest));
+            return Ok((rest,
+                       CallFrameInstruction::Offset {
+                register: register as u64,
+                factored_offset: offset,
+            }));
+        }
+
+        if high_bits == constants::DW_CFA_restore.0 {
+            let register = instruction & CFI_INSTRUCTION_LOW_BITS_MASK;
+            return Ok((rest, CallFrameInstruction::Restore { register: register as u64 }));
+        }
+
+        debug_assert!(high_bits == 0);
+
+        if instruction == constants::DW_CFA_nop.0 {
```

Comment:
> Nit: how about let instruction = DwCfa(instruction); and then drop the .0.

Done

> Is there a reason you used if statements instead of match?

I wanted to avoid rightward drift since the high bit masking would lead to two matches, each of which is double indentation, so quad indentation. I ended up making the second larger set (when the high bits aren't set) a match.